### ISSUE TITLE
Support for Graal 21.0

### ISSUE THREAD

**Issue by Kineolyan** [borkdude/clj-reflector-graal-java11-fix#3]
I see that GraalVM 21.0 has been released but that there are no corresponding version of the reflector patch.
Would it mean that the issue went away with this latest version? Or more likely, that a new version for this project has not been made yet.
In the latter case, is there something that I can do to help?

Cheers

**Comment by borkdude** [comment]
@Kineolyan GraalVM has new support for resolving `MethodHandle` which was the problem with previous versions of GraalVM and `clojure.lang.Reflector`. I have experimented with 21.0 and it seems the workaround from this library is no longer needed. If it turns out it is needed, I would be happy to make another release. Keep me posted.

**Comment by borkdude** [comment]
It seems this lib is no longer needed with GraalVM 21 onwards.

**Comment by sogaiu** [comment]
Thanks for having made it though :)

**Comment by Kineolyan** [comment]
My bad. I did not answer. Thanks for doing it. In my one and only test so far, I add issues compiling without your tool. Let me try again and come back to this, as it is very likely that the issue is on my side.

**Comment by borkdude** [comment]
I think instead of this lib, you will now have to add this to your [reflection.json](https://github.com/oracle/graal/blob/master/substratevm/Reflection.md) config:

```
[{"name": "java.lang.reflect.AccessibleObject",
  "methods" : [{"name":"canAccess"}]}
]
```

**Comment by Kineolyan** [comment]
That's possible. I also checked how you did it for clj-kondo [1].
For my project, I was using the lein plugin io.taylorwood/lein-native-image [2] for this, without any special reflection.json config.

I actually ran the build copying your options and it worked like a charm. To really go to the bottom of **my** issue, I look at more of your projects, as well as this plugin I use.

So I confirm this fix is not needed anymore :tada: and thank you again for making it possible until now to have native images for clojure apps.
Have a nice evening.

[1] https://github.com/clj-kondo/clj-kondo/blob/c562760c6d07cede8c7d7aadbe1a45593c26f360/script/compile
[2] https://github.com/taylorwood/lein-native-image

**Comment by borkdude** [comment]
@Kineolyan The commit you are referring to is about 20.1.0, not 21.0.0, but great that it worked :).

**Comment by Kineolyan** [comment]
Yes, I know. I mostly used the options from the compile script. I know that the rest is still using this project. I was mostly looking for a way to manually call `native-image`.

**Comment by WhittlesJr** [comment]
@borkdude , I'm trying to get this running on NixOS. I've updated to 0.2.9 and graalvm-ce 21.0.0:

https://github.com/WhittlesJr/nixpkgs/commit/8417e64524c20cbcda2c0071f4d3aa601c0abfd7

But despite using your reflection.json, which contains config for `java.lang.reflect.AccessibleObject.canAccess`,  I'm still getting:

```
----- Error --------------------------------------------------------------------
Type:     java.lang.NoSuchMethodError
Message:  java.lang.reflect.AccessibleObject.canAccess(java.lang.Object)
Location: 10:1

----- Stack trace --------------------------------------------------------------
environ.core/slurp-file    - environ/core.cljc:45:8
environ.core/read-env-file - environ/core.cljc:53:22
environ.core/read-env-file - environ/core.cljc:52:8
environ.core/merge-env     - environ/core.cljc:70:12
environ.core/merge-env     - environ/core.cljc:64:8
environ.core/read-env      - environ/core.cljc:69:3
environ.core/read-env      - environ/core.cljc:68:8
environ.core               - environ/core.cljc:81:7
```
Do you have any ideas?

**Comment by borkdude** [comment]
@WhittlesJr Which reflection.json are you actually using? The above config should do it.

**Comment by WhittlesJr** [comment]
@borkdude it's using this one, from your release page: https://github.com/babashka/babashka/releases/download/v0.2.9/babashka-0.2.9-reflection.json

It has the relevant config, so I figure it ought to work, but for whatever reason it's not working. Could it have something to do with the native-image invocation?

```
    native-image \
      -jar ${src} \
      -H:Name=bb \
      -H:+ReportExceptionStackTraces \
      -J-Dclojure.spec.skip-macros=true \
      -J-Dclojure.compiler.direct-linking=true \
      "-H:IncludeResources=BABASHKA_VERSION" \
      "-H:IncludeResources=SCI_VERSION" \
      "-H:ReflectionConfigurationFiles=${insert your reflection.json here}" \
      --initialize-at-build-time \
      -H:Log=registerResource: \
      -H:EnableURLProtocols=http,https \
      --enable-all-security-services \
      -H:+JNI \
      --verbose \
      --no-fallback \
      --no-server \
      --report-unsupported-elements-at-runtime \
      "--initialize-at-run-time=org.postgresql.sspi.SSPIClient" \
      "-J-Xmx4500m"
```

**Comment by borkdude** [comment]
@WhittlesJr That should work. Can you make a syntax error in the JSON to see if that makes the build fail? Just to see if the JSON file is picked up.

**Comment by WhittlesJr** [comment]
@borkdude , munging the config does cause it to throw: `Error: Error parsing reflection configuration` etc. It seems to be pulling it in...

**Comment by borkdude** [comment]
@WhittlesJr I don't know. I can't reproduce this on CI where I build for macOS, linux, linux static and Windows.

**Comment by WhittlesJr** [comment]
I'll try summoning the NixOS maintainers of the package here:
@DerGuteMoritz
@bandresen
@bhougland
@jlesquembre

**Comment by WhittlesJr** [comment]
Hold on, it seems like it was pointed at an old version of reflection.json? The URL was correct but the sha256 was old... I have no idea why that even worked. Everything is fine now. babashka is packaged, up-to-date, and working on NixOS.

**Comment by DerGuteMoritz** [comment]
> The URL was correct but the sha256 was old... I have no idea why that even worked.

Probably because you still had the old file matching the old checksum in your nix store. The URL itself does not factor into the store path. Glad you figured it out :pray:
